torch/distributed/_composable/fsdp/_fsdp_param.py

```diff
@@ -251,13 +287,40 @@ def init_all_gather_outputs(
         ]
 
     def init_unsharded_param(self):
-        if hasattr(self, "_unsharded_param"):
-            return  # already initialized
-        # For the default path (no post-all-gather), the all-gather output
-        # gives the unsharded parameter data directly
-        assert len(self.all_gather_outputs) == 1
+        if hasattr(self, "_unsharded_param"):  # after the 1st all-gather
```

Comment:
Might be similar to my previous comment, I feel this can be simplified by treating `torch.Tensor` and tensor subclass the same way, we can always check if the inner tensor have fsdp_pre_all_gather and fsdp_post_all_gather, if so we do additional handling, if not they are just vanilla tensor path?

Reply:
let me think through this more and see if we can unify

Reply:
I think that the code below is as unified as it can be 🤔 . The tail logic for constructing the unsharded parameter is shared.

The divergent part is required since for the extensions path we (1) unflatten all-gather outputs and (2) have to call `fsdp_post_all_gather` either with its out variant or not. There is no reason to unflatten all-gather outputs for the default `torch.Tensor` path, and it would be an extra view op.